### PR TITLE
Refactor fetch instance resource from API for error message generation

### PR DESCRIPTION
**Pull request by cflynn07**
Refactor to fetch instance object directly from API for error display information generation.
Co-PR: https://github.com/CodeNow/navi/pull/94
# 

Review:
- [x] @anandkumarpatel
- [x] @Nathan219

### CODE REVIEW

File: bin/www [hunk 1]
```diff
@@ -4,10 +4,22 @@
...
 var http = require('http');
 
+var app = require('../app');
+
+var requiredEnvKeys = [
+  'HELLO_RUNNABLE_GITHUB_TOKEN',
+  'API_HOST'
+];
+if (!hasKeypaths(process.env, requiredEnvKeys)) {
+  console.log('Missing required ENV keys');
```
Comment (anandkumarpatel) [inline bin/www]: Let's  use Bunyan here. Also error cat create and report error and exit on cb

File: app.js [hunk 2]
```diff
@@ -1,63 +1,219 @@
...
-        options.status = 'failed to build';
-      } else {
-        options.status = 'is ' + options.status;
-      }
+  }
+
+  if (req.query.type === 'signin') {
+    log.trace('processNaviError type signin');
+    return res.render('pages/signin', options);
+  } else if (req.query.type === 'not_running') {
```
Comment (anandkumarpatel) [inline app.js]: no need for `else if` your returning above

File: bin/www [hunk 3]
```diff
@@ -4,10 +4,22 @@
...
 var http = require('http');
 
+var app = require('../app');
+
+var requiredEnvKeys = [
+  'HELLO_RUNNABLE_GITHUB_TOKEN',
+  'API_HOSTNAME'
+];
+if (!hasKeypaths(process.env, requiredEnvKeys)) {
+  console.log('Missing required ENV keys');
```
Comment (anandkumarpatel) [inline bin/www]: loggly

File: bin/www [hunk 4]
```diff
@@ -4,10 +4,22 @@
...
+  console.log(requiredEnvKeys);
+  process.exit(1);
```
Comment (anandkumarpatel) [inline bin/www]: rollbar error

File: app.js [hunk 5]
```diff
@@ -1,63 +1,219 @@
...
+        break;
+      case 'unknown':
+        options.headerText = 'unknown';
+        res.render('pages/dead', options);
+        break;
     }
-    res.render('pages/' + page, options);
-  } else {
-    res.render('pages/invalid', options);
+  } else if (req.query.type === 'ports'){
```
Comment (anandkumarpatel) [inline app.js]: not really a fan of these long running `else if's` have to keep scrolling up to see which chain it is a part of. Would it be hard converting this into returns?

**Comment by anandkumarpatel** [comment]
overall logic seems good. might be time to break out express-app and app logic into 2 files? not required just a question.

File: app.js (hunk 5 above)
Comment (cflynn07) [inline app.js]: I converted it into returns

File: app.js [hunk 6]
```diff
@@ -1,63 +1,225 @@
...
 
 // view engine setup
 app.set('views', path.join(__dirname, 'views'));
 app.set('view engine', 'jade');
 
-var locals = {
-  version: version
+assign(app.locals, {
+  localVersionversion: version,
+  absoluteUrl: process.env.ABSOLUTE_URL || 'detention.runnable.io'
```
Comment (anandkumarpatel) [inline app.js]: remove `||` no reason to point to prod as default

File: bin/www [hunk 7]
```diff
@@ -4,10 +4,28 @@
...
+  'HELLO_RUNNABLE_GITHUB_TOKEN',
+  'API_HOSTNAME'
+];
+if (!hasKeypaths(process.env, requiredEnvKeys)) {
+  log.error({
+    requiredEnvKeys: requiredEnvKeys,
+    env: process.env
+  }, 'Missing required ENV keys');
+  // send to rollbar
+  ErrorCat.report(new Error('Detention missing required ENV values'));
```
Comment (anandkumarpatel) [inline bin/www]: this error will not get sent to rollbar, call it in the callback of this funtion
ErrorCat.prototype.report = function (err, req, cb) {

File: bin/www [hunk 8]
```diff
@@ -25,7 +43,9 @@ var server = http.createServer(app);
  * Listen on provided port, on all network interfaces.
  */
 
-server.listen(port);
+app.loginSuperUser(function () {
+  server.listen(port);
+});
 server.on('error', onError);
 server.on('listening', onListening);
```
Comment (anandkumarpatel) [inline bin/www:51]: where is `onListening` defined?

File: bin/www [hunk 9]
```diff
@@ -25,7 +43,9 @@ var server = http.createServer(app);
  * Listen on provided port, on all network interfaces.
  */
 
-server.listen(port);
+app.loginSuperUser(function () {
+  server.listen(port);
+});
 server.on('error', onError);
```
Comment (anandkumarpatel) [inline bin/www:50]: move this above app.login, was a bug before, if listen error sync it will not get called.

Comment (anandkumarpatel) [inline bin/www:50]: also where is `onError` ?!

File: bin/www (hunk 8 above)
Comment (cflynn07) [inline bin/www:51]: 20 lines down

Comment (anandkumarpatel) [inline bin/www:51]: ahh nm, git diff messed me up

File: bin/www (hunk 9 above)
Comment (anandkumarpatel) [inline bin/www:50]: ahh nm, git diff messed me up

File: logger.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+ * @module lib/logger
+ */
+'use strict';
+
+var bunyan = require('bunyan');
+var envIs = require('101/env-is');
+var path = require('path');
+var put = require('101/put');
+
+var serializers = put(bunyan.stdSerializers, {
```
Comment (anandkumarpatel) [inline logger.js]: nothing here?

File: logger.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  // put serializers here
+});
+
+var logger = bunyan.createLogger({
+  name: 'detention',
+  streams: [{
+    level: process.env.LOG_LEVEL_STDOUT || 'trace',
+    stream: process.stdout
+  }],
+  serializers: serializers,
```
Comment (anandkumarpatel) [inline logger.js]: just set this to `bunyan.stdSerializers`

File: logger.js (hunk 10 above)
Comment (anandkumarpatel) [inline logger.js]: remove unless you plan on adding something in the near future

File: bin/www (hunk 9 above)
Comment (cflynn07) [inline bin/www:50]: Was confused by what you thought you were seeing....

File: logger.js (hunk 10 above)
Comment (cflynn07) [inline logger.js]: I left it because yeah, I imagine it wont be long at all until I add something.

File: logger.js (hunk 11 above)
Comment (cflynn07) [inline logger.js]: sure will do

File: app.js [hunk 12]
```diff
@@ -1,63 +1,225 @@
...
+  });
+
+  if (req.instance) {
+    options.branchName = keypather.get(req.instance, 'attrs.contextVersion.branch');
+    // Temp missing pending resolution of SAN-3018
+    // https://runnable.atlassian.net/browse/SAN-3018
+    options.instanceName = keypather.get(req.instance, 'attrs.lowerName');
+    options.ownerName = keypather.get(req.instance, 'attrs.owner.username');
+    var ports = keypather.get(req.instance, 'attrs.container.ports');
+    options.ports = Object.keys(ports).map(function (portKey) {
```
Comment (Nathan219) [inline app.js]: if ports is undefined here, it causes an exception

File: app.js [hunk 13]
```diff
@@ -1,63 +1,225 @@
...
+];
 
 // view engine setup
 app.set('views', path.join(__dirname, 'views'));
 app.set('view engine', 'jade');
 
-var locals = {
-  version: version
+assign(app.locals, {
+  localVersionversion: version,
```
Comment (Nathan219) [inline app.js]: Possibly a grep bug?  Should just be localVersion

Comment (cflynn07) [inline app.js]: whoops yeah good catch

File: app.js [hunk 14]
```diff
@@ -1,63 +1,225 @@
...
+     *
+     * Userland hipache will only route to navi if a request is made to an elastic url on a port
+     * that's explicitly set on the instance (we set hipache redis entries when ports are exposed)
+     * otherwise userland-hipache will return an error page due to a lack of a redis entry.
+     *
+     * Probably could fix by patching Hipache or perhaps reading the manual to see if there's a
+     * some kind of forward-for-all-ports functionality
+     *
+     * Anand if you read this Monday morning lets chat about it at 3pm
+     */
```
Comment (Nathan219) [inline app.js:210]: Doesn't ports work right now?  Or have I only been seeing them due to errors?

Comment (cflynn07) [inline app.js:210]: The unresponsive page will show you a list of ports. That's probably what you've been seeing. We don't have anything yet that can display a message if you try to access a port that isn't set as exposed on the instance document. We basically need to hack hipache to get that.

File: app.js [hunk 15]
```diff
@@ -1,63 +1,231 @@
...
+
+  [
+    'redirectUrl',
+    'shortHash'
+  ].forEach(function (option) {
+    options[option] = req.query[option];
+  });
+
+  if (req.instance) {
+    options.branchName = keypather.get(req.instance, 'attrs.contextVersion.branch');
```
Comment (Nathan219) [inline app.js]: use req.instance.getBranchName() here

File: app.js [hunk 16]
```diff
@@ -1,63 +1,231 @@
...
+    // Temp missing pending resolution of SAN-3018
+    // https://runnable.atlassian.net/browse/SAN-3018
+    options.instanceName = keypather.get(req.instance, 'attrs.lowerName');
+    options.ownerName = keypather.get(req.instance, 'attrs.owner.username');
```
Comment (Nathan219) [inline app.js:132]: I found in my testing that this coming back null (using prod api).  I only get the githubId back

Comment (cflynn07) [inline app.js:132]: Yes @Nathan219 this JIRA ticket is tracking that.
https://runnable.atlassian.net/browse/SAN-3018

Not a blocker for deploying the refactored detention as is.

Comment (cflynn07) [inline app.js:132]: Good eye tho finding that.

File: app.js (hunk 15 above)
Comment (cflynn07) [inline app.js]: done 
https://github.com/CodeNow/detention/commit/4c6656c48dcc538ffe2c46861302f4033df98e38

File: app.js [hunk 17]
```diff
@@ -1,63 +1,231 @@
...
+  ].forEach(function (option) {
+    options[option] = req.query[option];
+  });
+
+  if (req.instance) {
+    options.branchName = req.instance.getBranchName();
+    // Temp missing pending resolution of SAN-3018
+    // https://runnable.atlassian.net/browse/SAN-3018
+    options.instanceName = keypather.get(req.instance, 'attrs.lowerName');
+    options.ownerName = keypather.get(req.instance, 'attrs.owner.username');
```
Comment (Nathan219) [inline app.js:132]: This is null, which makes the links (view logs, go to container) not work

File: README.md [hunk 18]
```diff
@@ -1,5 +1,10 @@
-# detention
-Handles our 404
+Detention
+=========
```
Comment (bkendall) [inline README.md]: blah... use `#` - one less line, and consistent for the rest of the readme...

File: app.js (hunk 17 above)
Comment (cflynn07) [inline app.js:132]: Did you see this? https://github.com/CodeNow/detention/pull/8#discussion_r46635696

Comment (Nathan219) [inline app.js:132]: That's weird that the comment wasn't on the full change...

**Comment by Nathan219** [comment]
LGTM

**Pull request merged**